Added integration test of secrets.

erictune opened the pull request:
Tests apiserver side functionality.

Also deleted dead code in auth_test.go.

erictune committed:
Added integration test of secrets.

Tests apiserver side functionality.

Also deleted dead code in auth_test.go.

mbforbes commented on test/integration/secret_test.go:
```diff
@@ -0,0 +1,188 @@
+// +build integration,!no-etcd
+
+/*
+Copyright 2014 Google Inc. All rights reserved.
```
2015

mbforbes commented on test/integration/secret_test.go:
```diff
@@ -0,0 +1,188 @@
+	}
+}
+
+// doTestSecrets test secrets for one api version.
```
**D**oTestSecrets

mbforbes commented on test/integration/secret_test.go:
```diff
@@ -0,0 +1,188 @@
+		},
+	}
+
+	// Create a pod to consume secret in its same namespace
```
Is OSS Go code also pedantic about ending comments with periods?

mbforbes commented on test/integration/secret_test.go:
```diff
@@ -0,0 +1,188 @@
+	// TODO: need a way to prevent pods from using arbitary secrets. (#4957)
+
+	// Can create pod referencing secret in same namespace which does not exist.
+	// This pod may fail to run, but we don't currently prevent this, and this test can't check
```
nit-to-end-all-nits: Line wrapping inconsistent, which slows down my brains' Go parsing speed

mbforbes commented on test/integration/secret_test.go:
```diff
@@ -0,0 +1,188 @@
+
+package integration
+
+// This file tests authentication and (soon) authorization of HTTP requests to a master object.
```
I think this comment is from `auth_test.go` and needs to be updated.

erictune commented:
all comments addressed.

mbforbes commented:
Thanks for addressing comments so fast.

I did a syntactic review; assigning Quinton (arbitrarily; feel free to re-assign) for a semantic one.

My one stab at a question: how would this test fail? Do the pods we're trying to schedule fail if the secret doesn't exist in their namespace?

erictune commented:
We are going to want them to fail validation, but they don't yet.  This was
my attempt at TDD.  Once I add the validation check, I'll make the second
and third test cases fail.

On Fri, Mar 20, 2015 at 3:37 PM, Maxwell Forbes <notifications@github.com>
wrote:

> Thanks for addressing comments so fast.
>
> I did a syntactic review; assigning Quinton (arbitrarily; feel free to
> re-assign) for a semantic one.
>
> My one stab at a question: how would this test fail? Do the pods we're
> trying to schedule fail if the secret doesn't exist in their namespace?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5686#issuecomment-84170866>
> .
>

quinton-hoole commented on test/integration/secret_test.go:
```diff
@@ -0,0 +1,185 @@
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
```
Do you need a deferred cleanup here?  No sure.

erictune replied:
I don't think so.

quinton-hoole replied:
You're right.

quinton-hoole commented on test/integration/secret_test.go:
```diff
@@ -0,0 +1,185 @@
+	testCases := []string{
+		"v1beta1",
+		"v1beta2",
+	}
```
This seems like it belongs at a higher scope.  How is it kept current?

erictune replied:
It is what other tests in this directory do.

quinton-hoole replied:
Fair enough.  Best open a bug to add v1beta3 across the board soon.

quinton-hoole commented on test/integration/secret_test.go:
```diff
@@ -0,0 +1,185 @@
+		},
+		{
+			ObjectMeta: api.ObjectMeta{
+				Name:      "secret",
```
Suggest factoring out duplicate literal "secret".

quinton-hoole commented on test/integration/secret_test.go:
```diff
@@ -0,0 +1,185 @@
+	for _, s := range secrets {
+		defer deleteSecretOrErrorf(t, client, s.Namespace, s.Name)
+		if _, err := client.Secrets(s.Namespace).Create(&s); err != nil {
+			t.Errorf("unable to create test secret %s: %v", s.Name, err)
```
You'll need the namespace in here too.  Consider using types.NamespacedName.

erictune replied:
fixed.

quinton-hoole commented on test/integration/secret_test.go:
```diff
@@ -0,0 +1,185 @@
+					Image: "fakeimage",
+					VolumeMounts: []api.VolumeMount{
+						{
+							Name:      "secvol",
```
Suggest factoring out duplicate literal "secvol".

quinton-hoole commented on test/integration/secret_test.go:
```diff
@@ -0,0 +1,185 @@
+						Secret: &api.SecretVolumeSource{
+							Target: api.ObjectReference{
+								Kind:      "Secret",
+								Namespace: "XXX",
```
I assume that this "XXX" literal and the one above are not required to be equal?  If so, suggest making them non-equal to clarify that there's no relationship between the two.

erictune replied:
no longer relevant.

quinton-hoole commented on test/integration/secret_test.go:
```diff
@@ -0,0 +1,185 @@
+	defer deletePodOrErrorf(t, client, ns1, pod.Name)
+	if _, err := client.Pods(ns1).Create(pod); err != nil {
+		t.Errorf("Failed to create pod: %v", err)
+	}
```
Suggest factoring out this cut'n pasted code.  And to illustrate why this is a good idea, I think you have bug due to cut'n'paste in defer deletePod... line.

quinton-hoole replied:
Actually... maybe not.  Ignore me for now.

quinton-hoole replied:
Good idea to factor out this common code into a separate function.  It will make this test shorter and easier to follow.

erictune replied:
I think copy-paste code is more desirable in tests than in regular code.
Here is why:
http://googletesting.blogspot.com/2014/07/testing-on-toilet-dont-put-logic-in.html

quinton-hoole replied:
Fair enough, I can understand that argument.  I'm still hurting from having to fix/rewrite a bunch of tests that had tons of cut-n-pasted bugs throughout them.   LGTM either way :-)

quinton-hoole commented on test/integration/secret_test.go:
```diff
@@ -0,0 +1,185 @@
+	// Create a pod to consume secret in its same namespace.
+	pod.ObjectMeta.Name = "uses-same-ns-secret"
+	pod.Spec.Volumes[0].VolumeSource.Secret.Target.Namespace = ns1
+	defer deletePodOrErrorf(t, client, ns1, pod.Name)
```
Deferred deletion should come after successful creation?

erictune replied:
fixed.

quinton-hoole commented on test/integration/secret_test.go:
```diff
@@ -0,0 +1,185 @@
+	// Can create pod referencing secret in same namespace which does not exist.
+	// This pod may fail to run, but we don't currently prevent this, and this
+	// test can't check whether the kubelet actually pulls the secret.
+	pod.ObjectMeta.Name = "uses-non-existent--secret"
```
Did you intend a double "--" here, or is it just a typo?

erictune replied:
fixed

quinton-hoole commented:
Ping!  Eric, I assume that this is waiting for you, and not me?

erictune commented:
I've rewritten the test after rebasing to pick up #5807

quinton-hoole commented:
LGTM.  Can be merged as soon as Travis goes green.

erictune commented:
green
